### PR TITLE
Book store

### PR DESCRIPTION
**Pull request by NoelLincoln**
Hello 👋

My coding partner and I worked on the awesome books project using javascript objects as required :
- Users can add books
- The page displays added  books
- Added books are saved to local storage
- Users can delete books
- Deleted books are removed from the page and from local storage.

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+document.addEventListener('DOMContentLoaded', () => {
+  const bookTitle = document.getElementById('title');
+  const bookAuthor = document.getElementById('author');
+  const article = document.getElementById('bookList');
+
+  const addButton = document.getElementById('add');
+  const saveItemToLocalStorage = (title, author) => {
+    const savedBooks = JSON.parse(localStorage.getItem('savedBooks')) || [];
+    savedBooks.push({ title, author });
```
Comment (shahadat3669) [inline script.js]: Well done on your implementation! However, I'd like to suggest a slight improvement that could enhance your code quality.
- [ ] [Optional]  Currently, you are using the `array.push()` method to add new elements to an array. While this approach works well, I recommend considering the use of the `array.concat()` method instead. The [concat()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat) method of [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array) instances allows you to merge two or more arrays. The key advantage here is that it doesn't modify the existing arrays; instead, it returns a new array that combines the contents of the original arrays. This can help maintain data [immutability](https://developer.mozilla.org/en-US/docs/Glossary/Immutable) and prevent unexpected side effects.

By using `array.concat()` in your context, you'll be creating a new array that includes your newly generated book while keeping the original array intact. This can contribute to a more robust and maintainable codebase. Keep up the excellent work, and happy coding! 🌟📝🚀"

File: script.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  const removeItemFromLocalStorage = (div) => {
+    const savedBooks = JSON.parse(localStorage.getItem('savedBooks')) || [];
+    const title = div.querySelector('p:first-child').innerHTML;
+    const author = div.querySelector('p:nth-child(2)').innerHTML;
+    const index = savedBooks.findIndex(
+      (item) => item.title === title && item.author === author,
+    );
+    if (index !== -1) {
+      savedBooks.splice(index, 1);
```
Comment (shahadat3669) [inline script.js]: Well done on your implementation! However, I'd like to suggest a small change that could enhance your code quality.

- [ ] [Optional]  Currently, you are using the `array.splice()` method to remove elements from the array. While this approach works as intended, I recommend considering the use of the `array.slice()` method instead. The [slice()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice) method of [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array) instances returns a [shallow copy](https://developer.mozilla.org/en-US/docs/Glossary/Shallow_copy) of a portion of an array, selected from the start index to the end index (end index not included). This new array copy includes the items within the specified range, leaving the original array untouched. The key advantage of using `array.slice()` for this purpose is that it maintains the immutability of the original array. This means that the removal of a book using `array.slice()` won't modify the existing array but will return a new array with the desired items removed. This aligns well with the principles of data immutability and can lead to more predictable and maintainable code.

In summary, while both `array.splice()` and `array.slice()` are valid methods for removing elements, the choice to use `array.slice()` can contribute to clearer code that adheres to the principles of immutability. Keep up the excellent work, and happy coding! 🌟📝🚀

File: script.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      localStorage.setItem('savedBooks', JSON.stringify(savedBooks));
+    }
+  };
```
Comment (shahadat3669) [inline script.js]: Well, your remove function is functioning slightly as expected, allowing users to delete books. However, there are a couple of adjustments that can enhance its functionality:

- [ ] When there are two or more books with the same title and author, the current implementation sometimes deletes the wrong book from local storage. For instance, if you attempt to delete the second book, it might remove the first one's data from local storage. To address this, I recommend implementing a more precise way to identify and delete the correct book from local storage.

[Awesome-books-one.webm](https://github.com/NoelLincoln/awesome-books/assets/55840999/d92f315d-39bc-4d1d-98a7-d96e167a989e)

- [ ] [Optional] To make the removal process more accurate and efficient, I suggest adding a unique identifier (like an `id`) to each book object. This identifier should be distinct for every book. You can then use this unique `id` to precisely locate and delete the desired book from both the DOM and local storage. This approach ensures accurate book removal and maintains the integrity of your local storage data.

By incorporating these adjustments, you'll enhance the accuracy and reliability of your book removal functionality. Keep up the good work and strive for a seamless user experience! 📖🗑️🚀

File: script.js
Comment (shahadat3669) [inline script.js:1]: - [ ] Well, I noticed that when a user adds a book, you're correctly saving it to the local storage. However, there's a small issue - upon reloading the page, the books that were added and saved in the local storage are not being displayed on the webpage. According to the project requirements, the webpage should show the books that are present in the local storage even after a page reload. To address this, I recommend updating your code so that upon reloading the page, you retrieve the books stored in the local storage and display them on the webpage. This way, users can see the existing books even after navigating away from the page. It's an important feature to ensure a seamless user experience and meet the project specifications. Keep up the good work and ensure your application functions smoothly both during and after page reloads! 📚💻🔄🚀

**Review by shahadat3669 — CHANGES_REQUESTED** [review]
Hi @NoelLincoln, @MPDADDY, 

Great job so far! 👏 You've done well, and you're almost there for the next project ⏭️, but there are a few issues 🐛 that you still need to work on.




## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ has a function that adds a new book to the collection.
✅ displays all books.
✅ clicks the "Add" button, a new book is added to the collection

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadata3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by NoelLincoln** [comment]
The requested changes have been implemented

**Review by shahadat3669 — APPROVED** [review]
Hi @NoelLincoln, @MPDADDY, 

Your project is complete!  There is nothing else to say other than it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
✅ Good use of `array.concat()` method
✅ Good use of `array.slice()` method
✅ God use of unique ID

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table  width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadata3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
